File: linkedinsdk/src/main/java/com/AbdAllahAbdElFattah13/linkedinsdk/linkedin_builder/LinkedInBuilder.java
package com.AbdAllahAbdElFattah13.linkedinsdk.linkedin_builder;

import android.content.Context;
import android.content.Intent;
import android.util.Log;

import androidx.annotation.NonNull;

import com.AbdAllahAbdElFattah13.linkedinsdk.LinkedInAuthenticationActivity;
import com.AbdAllahAbdElFattah13.linkedinsdk.RetrieveBasicProfileAsyncTask;
import com.AbdAllahAbdElFattah13.linkedinsdk.helpers.OnBasicProfileListener;

import java.util.Random;

/**
 * Created by AbdAllah AbdElfattah on 11/04/2020,
 * Cairo, Egypt.
 */
public abstract class LinkedInBuilder {

    private Intent intent;
    private String state;

    public static final String CLIENT_ID = "client_id";
    public static final String CLIENT_SECRET_KEY = "client_secret";
    public static final String REDIRECT_URI = "redirect_uri";
    public static final String STATE = "state";

    public static final String TAG = "LinkedInAuth";


    public static final int ERROR_USER_DENIED = 11;
    public static final int ERROR_FAILED = 12;

    abstract LinkedInBuilder self();

    abstract void startActivityForResult(Intent intent, int requestCode);


    LinkedInBuilder(Context context) {
        this.intent = new Intent(context, LinkedInAuthenticationActivity.class);
    }

    public LinkedInBuilder setClientID(String clientID) {
        updateIntent(CLIENT_ID, clientID);
        return self();
    }

    public LinkedInBuilder setClientSecret(String clientSecret) {
        updateIntent(CLIENT_SECRET_KEY, clientSecret);
        return self();
    }

    public LinkedInBuilder setRedirectURI(String redirectURI) {
        updateIntent(REDIRECT_URI, redirectURI);
        return self();
    }

    public LinkedInBuilder setState(String state) {
        this.state = state;
        updateIntent(STATE, state);
        return self();
    }

    public void authenticate(int requestCode) {
        if (validateAuthenticationParams()) {
            if (state == null) {
                generateState();
            }
            this.startActivityForResult(intent, requestCode);
        }
    }

    private void updateIntent(String key, String value) {
        intent.putExtra(key, value);
    }

    private boolean validateAuthenticationParams() {

        if (intent.getStringExtra(CLIENT_ID) == null) {
            Log.e(TAG, "Client ID is required", new IllegalArgumentException());
            return false;
        }

        if (intent.getStringExtra(CLIENT_SECRET_KEY) == null) {
            Log.e(TAG, "Client Secret is required", new IllegalArgumentException());
            return false;
        }

        if (intent.getStringExtra(REDIRECT_URI) == null) {
            Log.e(TAG, "Redirect URI is required", new IllegalArgumentException());
            return false;
        }

        return true;
    }

    private void generateState() {
        String ALLOWED_CHARACTERS = "0123456789qwertyuiopasdfghjklzxcvbnmMNBVCXZLKJHGFDSAQWERTYUIOP";
        final Random random = new Random();
        final StringBuilder sb = new StringBuilder(16);
        for (int i = 0; i < 16; ++i)
            sb.append(ALLOWED_CHARACTERS.charAt(random.nextInt(ALLOWED_CHARACTERS.length())));
        this.state = sb.toString();
        intent.putExtra(STATE, state);
    }

    public static void retrieveBasicProfile(@NonNull String accessToken, long accessTokenExpiry, @NonNull OnBasicProfileListener onBasicProfileListener) {
        new RetrieveBasicProfileAsyncTask(accessToken, accessTokenExpiry, onBasicProfileListener).execute();
    }

}

File: linkedinsdk/src/main/java/com/AbdAllahAbdElFattah13/linkedinsdk/ui/linkedin_builder/LinkedInBuilder.kt
package com.AbdAllahAbdElFattah13.linkedinsdk.ui.linkedin_builder

import android.content.Context
import android.content.Intent
import android.util.Log
import com.AbdAllahAbdElFattah13.linkedinsdk.ui.LinkedInAuthenticationActivity
import java.util.*

/**
 * Created by AbdAllah AbdElfattah on 11/04/2020,
 * Cairo, Egypt.
 */
abstract class LinkedInBuilder internal constructor(context: Context?) {

    private val intent: Intent = Intent(context, LinkedInAuthenticationActivity::class.java)
    private var state: String? = null
    private var accessTokenRetrievalOnlyRequest = false

    abstract fun self(): LinkedInBuilder
    abstract fun startActivityForResult(intent: Intent, requestCode: Int)

    fun setClientID(clientID: String): LinkedInBuilder {
        updateIntent(CLIENT_ID, clientID)
        return self()
    }

    fun setClientSecret(clientSecret: String): LinkedInBuilder {
        updateIntent(CLIENT_SECRET_KEY, clientSecret)
        return self()
    }

    fun setRedirectURI(redirectURI: String): LinkedInBuilder {
        updateIntent(REDIRECT_URI, redirectURI)
        return self()
    }

    fun setState(state: String): LinkedInBuilder {
        this.state = state
        updateIntent(STATE, state)
        return self()
    }

    fun setAccessTokenRetrievalOnlyRequest(accessTokenRetrievalOnlyRequest: Boolean): LinkedInBuilder {
        this.accessTokenRetrievalOnlyRequest = accessTokenRetrievalOnlyRequest
        intent.putExtra(ACCESS_TOKEN_ONLY, accessTokenRetrievalOnlyRequest)
        return self()
    }

    fun authenticate(requestCode: Int) {
        if (validateAuthenticationParams()) {
            if (state == null) {
                generateState()
            }
            startActivityForResult(intent, requestCode)
        }
    }

    private fun updateIntent(key: String, value: String) {
        intent.putExtra(key, value)
    }

    private fun validateAuthenticationParams(): Boolean {
        if (intent.getStringExtra(CLIENT_ID) == null) {
            Log.e(TAG, "Client ID is required", IllegalArgumentException())
            return false
        }
        if (intent.getStringExtra(CLIENT_SECRET_KEY) == null) {
            Log.e(TAG, "Client Secret is required", IllegalArgumentException())
            return false
        }
        if (intent.getStringExtra(REDIRECT_URI) == null) {
            Log.e(TAG, "Redirect URI is required", IllegalArgumentException())
            return false
        }
        return true
    }

    private fun generateState() {
        val allowedCharSet = "0123456789qwertyuiopasdfghjklzxcvbnmMNBVCXZLKJHGFDSAQWERTYUIOP"
        val random = Random()
        val sb = StringBuilder(16)
        for (i in 0..15) sb.append(allowedCharSet[random.nextInt(allowedCharSet.length)])
        state = sb.toString()
        intent.putExtra(STATE, state)
    }

    companion object {
        const val CLIENT_ID = "client_id"
        const val CLIENT_SECRET_KEY = "client_secret"
        const val REDIRECT_URI = "redirect_uri"
        const val STATE = "state"
        const val ACCESS_TOKEN_ONLY = "access_token_only"
        const val TAG = "LinkedInAuth"
        const val ERROR_USER_DENIED = 11
        const val ERROR_FAILED = 12
    }
}